borsh/tests/test_simple_structs.rs

```diff
@@ -217,3 +218,122 @@ fn test_simple_struct() {
     assert_eq!(decoded_f2.aa.len(), 2);
     assert!(decoded_f2.aa.iter().all(|f2_a| f2_a == &expected_a));
 }
+
+// sequence, no unit enums
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Eq, Clone, Copy, Debug)]
+#[borsh(use_discriminant = true)]
+#[repr(u16)]
+enum XY {
+    A,
+    B = 20,
+    C,
+    D(u32, u32),
+    E = 10,
+    F(u64),
+}
+
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Eq, Clone, Copy, Debug)]
+#[borsh(use_discriminant = false)]
+#[repr(u16)]
+enum XYNoDiscriminant {
+    A,
+    B = 20,
+    C,
+    D(u32, u32),
+    E = 10,
+    F(u64),
+}
+
+#[test]
+fn test_discriminant_serde_no_unit_type() {
+    let values = vec![XY::A, XY::B, XY::C, XY::E, XY::D(12, 14), XY::F(35325423)];
+    let expected_discriminants = [0u8, 20, 21, 10, 22, 11];
+
+    for (ind, value) in values.iter().enumerate() {
+        let data = value.try_to_vec().unwrap();
+        assert_eq!(data[0], expected_discriminants[ind]);
+        assert_eq!(from_slice::<XY>(&data).unwrap(), values[ind]);
+    }
+}
+
+#[test]
+fn test_discriminant_serde_no_unit_type_no_use_discriminant() {
+    let values = vec![
+        XYNoDiscriminant::A,
+        XYNoDiscriminant::B,
+        XYNoDiscriminant::C,
+        XYNoDiscriminant::D(12, 14),
+        XYNoDiscriminant::E,
+        XYNoDiscriminant::F(35325423),
+    ];
+    let expected_discriminants = [0u8, 1, 2, 3, 4, 5];
+
+    for (ind, value) in values.iter().enumerate() {
+        let data = value.try_to_vec().unwrap();
+        assert_eq!(data[0], expected_discriminants[ind]);
+        assert_eq!(from_slice::<XYNoDiscriminant>(&data).unwrap(), values[ind]);
+    }
+}
+
+// minimal
+#[derive(BorshSerialize)]
+#[borsh(use_discriminant = true)]
+enum MyDiscriminantEnum {
+    A = 20,
+}
+
+#[derive(BorshSerialize)]
+#[borsh(use_discriminant = false)]
+enum MyDiscriminantEnumFalse {
+    A = 20,
+}
+
+#[derive(BorshSerialize)]
+enum MyEnumNoDiscriminant {
+    A,
+}
+#[test]
+fn test_discriminant_minimal_true() {
+    assert_eq!(MyDiscriminantEnum::A as u8, 20);
+    assert_eq!(MyDiscriminantEnum::A.try_to_vec().unwrap(), vec![20]);
+}
+
+#[test]
+fn test_discriminant_minimal_false() {
+    assert_eq!(MyDiscriminantEnumFalse::A as u8, 20);
+    assert_eq!(
+        MyEnumNoDiscriminant::A.try_to_vec().unwrap(),
+        MyDiscriminantEnumFalse::A.try_to_vec().unwrap(),
+    );
+    assert_eq!(MyDiscriminantEnumFalse::A.try_to_vec().unwrap(), vec![0]);
+}
+
+// sequence
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Eq, Clone, Copy, Debug)]
+#[borsh(use_discriminant = false)]
+enum XNoDiscriminant {
+    A,
+    B = 20,
+    C,
+    D,
+    E = 10,
+    F,
+}
+
```

Comment:
extract all tests, related to enum discriminants, (added in pr and existing before) from this file `borsh/tests/test_simple_structs.rs` to a separate file  `borsh/tests/test_enum_discriminants.rs`